src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -434,12 +443,23 @@ func isCsrApproved(csr *certificatesv1.CertificateSigningRequest) bool {
 	return false
 }
 
-func (c controller) PostInstallConfigs(ctx context.Context, wg *sync.WaitGroup) {
+func (c *controller) updateFinalizingProgress(ctx context.Context, stage models.FinalizingStage) {
+	_ = utils.RetryWithContext(ctx, math.MaxInt, GeneralWaitInterval, c.log, func() error {
+		return c.ic.UpdateFinalizingProgress(ctx, c.ClusterID, stage)
+	})
+}
+
+func (c *controller) waitOnStage(ctx context.Context, stage models.FinalizingStage, interval time.Duration, predicate func() bool) error {
+	c.updateFinalizingProgress(ctx, stage)
+	return utils.WaitForeverForPredicate(ctx, interval, predicate)
+}
+
+func (c *controller) PostInstallConfigs(ctx context.Context, wg *sync.WaitGroup) {
 	defer func() {
 		c.log.Infof("Finished PostInstallConfigs")
 		wg.Done()
 	}()
-	err := utils.WaitForPredicateWithContext(ctx, LongWaitTimeout, GeneralWaitInterval, func() bool {
+	err := c.waitOnStage(ctx, models.FinalizingStageWaitingForFinalizing, GeneralWaitInterval, func() bool {
```

Comment:
This code runs in parallel with WaitAndUpdateNodesStatus as far as i remember.
From my point we don't need stage that says that we wait for stage. 
Can you explain why we actually need it?

Reply:
It might a bit technical, but timeouts are set on stages.  The assisted service needs to know that the controller is waiting on this stage.  This stage might be timed out if finalizing stage doesn't start after the timeout duration.
BTW, WaitAndUpdateNodesStatus might continue after finalizing started if there are some workers that are not done.

Reply:
I mean that assisted service can timeout on Finalizing stage itself, why it requires FinalizingStageWaitingForFinalizing ? Still not sure i understand, we don't have timeout on finalizing stage anymore? 
Still believe this stage is not needed and we don't need to timeout on it, we should timeout on regular finalizing stage and here to wait for cluster to go to finalizing stage. If we will get timeout cause 3 masters didn't join the cluster, cluster will go to error and context will be cancelled from main in all other cases we just should wait here.